### ISSUE TITLE
Garbage collector does not meet the specified time limit

### ISSUE THREAD

**Issue by SanyaShanin** [YoYoGames/GameMaker-Bugs#8268]
### Description

We actively use structures in our project, it is quite large, contains a lot of graphics and sounds. But we ran into a problem: the running game freezes for a fraction of a second stably every 15-20 seconds. Looking at the time schedule of the garbage collector, there is clearly a similarity between the peaks on the graph and the moments when the game freezes. We used the gc_target_frame_time function with different values, from 100 to 2000 microseconds, but the garbage collector still makes the game hang, instead of distributing the load between frames
![Image](https://github.com/user-attachments/assets/774f96f2-e99f-456b-b32d-7b78c250d66b)
![Image](https://github.com/user-attachments/assets/572e0f69-e57b-4c5a-b98b-64ce911ba530)
I hope there is something that can help, besides "don't use structures so often". Of course, we will work to reduce the number of structures being created, but this is a job that takes time. Initially, there is no warning in the help that the garbage collector may reduce performance when creating textures frequently. If this is not solvable, it would be nice to mention it in the documentation.

[https://api.gamemaker.io/api/github/downloads/73952591-6e9f-4111-8d36-700d30f227fa](url)

### Expected Change

Can you fix this, or explain what we need to do to make the garbage collector work more smoothly?

### Steps To Reproduce

this code can reproduce the problem we have:
`
CREATE:
frames_timer = 0;
parent_garbarge = {};
next_garbarge = parent_garbarge;

STEP:
frames_timer+=1;
if (frames_timer > 250) {
	frames_timer = 0;
	parent_garbarge = {};
	next_garbarge = parent_garbarge;
}
repeat 2000 {
	var old = next_garbarge;
	next_garbarge = {}
	old.child = next_garbarge;
}
`
I used it in the sample test project in the Camera object

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.8.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Sample Package Attached?

- [x] I have attached my bug package

### Sample Project Added?

- [x] I have included a small sample project

**Comment by SanyaShanin** [comment]
I used latest beta and 2024.8 version for testing. Results are same

**Comment by L-Clouds** [comment]
The documentation _does_ mention: "Please note that this is simply a _target_ value, as some operations still need to run completely in one frame and may sometimes take longer than the target time."

Whether your situation is one of those cases, however, I don't know.

**Comment by SanyaShanin** [comment]
> The documentation _does_ mention: "Please note that this is simply a _target_ value, as some operations still need to run completely in one frame and may sometimes take longer than the target time."
> 
> Whether your situation is one of those cases, however, I don't know.

Indeed, you are right. But there is no mention that due to the large amount of garbage, the collector can create freezes for 0.1-0.3 seconds. And more. For people without experience working with structures and the garbage collector, this can become a headache at the final stage of creating a game.

**Comment by danieltjewett** [comment]
The problem is that every so often, garbage collection iterates over every single generation 2 object, taking huge amounts of traversal times:
![Image](https://github.com/user-attachments/assets/6d74fd19-66bc-43fc-88b0-0fe84ad9368e)
I'm really hoping the problem here is that `gc_target_frame_time()` isn't being respected.  Any game with tons of instances, with methods, arrays and structs will have no way to circumvent this.  We're trying to convert things to lists and maps, but this is really not ideal.  And worse yet, rewriting the instance methods is almost a rewrite of our game.

**Comment by rwkay** [comment]
Lists and Maps still have to be traversed for Garbage Collection so I doubt that you are going to save anything by doing that. The individual maps and lists will not be garbage collected but the contents still need to be traversed to establish what is alive.

A more effective strategy would be to look at pooling structures and arrays that are often being reused - so they are not collected - and limiting creation.

using structs and arrays is not bad but frequently creating something transitively and then throwing it away is - keep an array/list of free'd structs and removing it from that rather than creating it would help more.

**Comment by danieltjewett** [comment]
@rwkay Pooling structures and arrays is probably a good solution, and might be better than rewriting everything with lists and maps.  However, how would one pool functions?
Even with a pooling system, the title "`gc_target_frame_time()` does not distribute the load to meet the specified time limit" I think still holds true and would still help alleviate much of the problems.  Is this issue being looked at and/or addressed?
